presto-native-execution/presto_cpp/main/operators/StorageBasedExchangeSource.h

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "velox/common/file/File.h"
+#include "velox/common/file/FileSystems.h"
+#include "velox/core/PlanNode.h"
+#include "velox/exec/Exchange.h"
+#include "velox/exec/Operator.h"
+#include "velox/serializers/PrestoSerializer.h"
+
+namespace facebook::presto::operators {
+
+class StorageBasedExchangeSource : public velox::exec::ExchangeSource {
+ public:
+  StorageBasedExchangeSource(
+      const std::string& taskId,
+      int destination,
+      std::shared_ptr<velox::exec::ExchangeQueue> queue,
+      const std::vector<std::string>& inputFolders,
+      std::shared_ptr<velox::filesystems::FileSystem> fileSystem,
+      velox::memory::MemoryPool* pool)
+      : ExchangeSource(taskId, destination, queue, pool) {
+    for (auto& folder: inputFolders) {
+      auto files = fileSystem->list(folder);
```

Comment:
Do we expect all files in a given directory be the "right" files? What happens if we run duplicate tasks (for speculative execution)? Do we need a way to exclude the duplicate files?